### ISSUE TITLE
DOMParser() leaks memory

### ISSUE THREAD

**Issue by pgmmpk** [jsdom/jsdom#2722]
### Basic info:

- **Node.js version:**  10.16.3
- **jsdom version:** 13.2.0

### Minimal reproduction case

```js

const { JSDOM } = require('jsdom');
const dom = new JSDOM(`<!DOCTYPE html><body></body>`);
const DOMParser = dom.window.DOMParser;

const xmltext = `<?xml version="1.0"?>
<catalog>
   <book id="bk101">
      <author>Gambardella, Matthew</author>
      <title>XML Developer's Guide</title>
      <genre>Computer</genre>
      <price>44.95</price>
      <publish_date>2000-10-01</publish_date>
      <description>An in-depth look at creating applications 
      with XML.</description>
   </book>
   <book id="bk102">
      <author>Ralls, Kim</author>
      <title>Midnight Rain</title>
      <genre>Fantasy</genre>
      <price>5.95</price>
      <publish_date>2000-12-16</publish_date>
      <description>A former architect battles corporate zombies, 
      an evil sorceress, and her own childhood to become queen 
      of the world.</description>
   </book>
   <book id="bk103">
      <author>Corets, Eva</author>
      <title>Maeve Ascendant</title>
      <genre>Fantasy</genre>
      <price>5.95</price>
      <publish_date>2000-11-17</publish_date>
      <description>After the collapse of a nanotechnology 
      society in England, the young survivors lay the 
      foundation for a new society.</description>
   </book>
   <book id="bk104">
      <author>Corets, Eva</author>
      <title>Oberon's Legacy</title>
      <genre>Fantasy</genre>
      <price>5.95</price>
      <publish_date>2001-03-10</publish_date>
      <description>In post-apocalypse England, the mysterious 
      agent known only as Oberon helps to create a new life 
      for the inhabitants of London. Sequel to Maeve 
      Ascendant.</description>
   </book>
   <book id="bk105">
      <author>Corets, Eva</author>
      <title>The Sundered Grail</title>
      <genre>Fantasy</genre>
      <price>5.95</price>
      <publish_date>2001-09-10</publish_date>
      <description>The two daughters of Maeve, half-sisters, 
      battle one another for control of England. Sequel to 
      Oberon's Legacy.</description>
   </book>
   <book id="bk106">
      <author>Randall, Cynthia</author>
      <title>Lover Birds</title>
      <genre>Romance</genre>
      <price>4.95</price>
      <publish_date>2000-09-02</publish_date>
      <description>When Carla meets Paul at an ornithology 
      conference, tempers fly as feathers get ruffled.</description>
   </book>
   <book id="bk107">
      <author>Thurman, Paula</author>
      <title>Splish Splash</title>
      <genre>Romance</genre>
      <price>4.95</price>
      <publish_date>2000-11-02</publish_date>
      <description>A deep sea diver finds true love twenty 
      thousand leagues beneath the sea.</description>
   </book>
   <book id="bk108">
      <author>Knorr, Stefan</author>
      <title>Creepy Crawlies</title>
      <genre>Horror</genre>
      <price>4.95</price>
      <publish_date>2000-12-06</publish_date>
      <description>An anthology of horror stories about roaches,
      centipedes, scorpions  and other insects.</description>
   </book>
   <book id="bk109">
      <author>Kress, Peter</author>
      <title>Paradox Lost</title>
      <genre>Science Fiction</genre>
      <price>6.95</price>
      <publish_date>2000-11-02</publish_date>
      <description>After an inadvertant trip through a Heisenberg
      Uncertainty Device, James Salway discovers the problems 
      of being quantum.</description>
   </book>
   <book id="bk110">
      <author>O'Brien, Tim</author>
      <title>Microsoft .NET: The Programming Bible</title>
      <genre>Computer</genre>
      <price>36.95</price>
      <publish_date>2000-12-09</publish_date>
      <description>Microsoft's .NET initiative is explored in 
      detail in this deep programmer's reference.</description>
   </book>
   <book id="bk111">
      <author>O'Brien, Tim</author>
      <title>MSXML3: A Comprehensive Guide</title>
      <genre>Computer</genre>
      <price>36.95</price>
      <publish_date>2000-12-01</publish_date>
      <description>The Microsoft MSXML3 parser is covered in 
      detail, with attention to XML DOM interfaces, XSLT processing, 
      SAX and more.</description>
   </book>
   <book id="bk112">
      <author>Galos, Mike</author>
      <title>Visual Studio 7: A Comprehensive Guide</title>
      <genre>Computer</genre>
      <price>49.95</price>
      <publish_date>2001-04-16</publish_date>
      <description>Microsoft Visual Studio 7 is explored in depth,
      looking at how Visual Basic, Visual C++, C#, and ASP+ are 
      integrated into a comprehensive development 
      environment.</description>
   </book>
</catalog>`

for (let i = 0; i < 100000; i++) {
    new DOMParser().parseFromString(xmltext, 'application/xml');
    if (i % 1000 == 0) {
        global.gc();
        const used = process.memoryUsage().heapUsed / 1024 / 1024;
        console.log(i, used);
    }
}
```

Result:
```
node --expose-gc cc.js
0 19.56842803955078
1000 251.05691528320312
2000 482.4790725708008
3000 713.8462448120117
4000 945.1195602416992
5000 1176.2345962524414

<--- Last few GCs --->

[64714:0x102808000]    22420 ms: Scavenge 1388.8 (1423.2) -> 1388.3 (1424.2) MB, 3.6 / 0.0 ms  (average mu = 0.114, current mu = 0.067) allocation failure 
[64714:0x102808000]    22427 ms: Scavenge 1389.1 (1424.2) -> 1388.6 (1424.7) MB, 4.2 / 0.0 ms  (average mu = 0.114, current mu = 0.067) allocation failure 
[64714:0x102808000]    22435 ms: Scavenge 1389.3 (1424.7) -> 1388.8 (1425.2) MB, 4.3 / 0.0 ms  (average mu = 0.114, current mu = 0.067) allocation failure 
<--- JS stacktrace --->
==== JS stack trace =========================================
...
FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
```

### How does similar code behave in browsers?

Works fine in Chrome (does not crash, no slowness).

**Comment by pmdartus** [comment]
Thanks for reporting this issue. I was able to reproduce the leak locally on `13.2.0` but not on `15.2.1` (latest version of jsdom). When digging a little further, it appears that the leak has been solved between with the `14.1.0` release. I am highly suspecting that the leak has been fixed with this commit https://github.com/jsdom/jsdom/commit/49353e224b6c59e48c3cc7da3700dbe93a2286ac.

@pgmmpk Could you check if you are also able to reproduce the memory leak on the latest version of jsdom?

**Comment by pgmmpk** [comment]
Confirm that problem is fixed in `15.2.1`. Thank you!
